examples/benchmarks/other/unify.effekt

```diff
@@ -0,0 +1,172 @@
+  case Var(name) =>
+    val substitutedType = subst.getOrElse(name) { () => ty }
+    subst.apply(substitutedType)
+  case Con(name, args) =>
+    Con(name, args.map { arg => subst.apply(arg) })
+}
+
+def unify(ty1: Type, ty2: Type, subst: Substitution): Substitution / Exception[UnificationError] = {
+  val substTy1 = subst.apply(ty1)
+  val substTy2 = subst.apply(ty2)
+
+  (substTy1, substTy2) match {
+    // If both are the same variable, return current substitution
+    case (Var(x), Var(y)) and x == y =>
+      subst
+
+    // If first is a variable, try to bind it
+    case (Var(x), _) =>
+      if (x.occurs(substTy2)) {
+        do raise(OccursCheckFailure(x, substTy2), "")
+      } else {
+        subst.put(x, substTy2)
+      }
+
+    // If second is a variable, try to bind it
+    case (_, Var(y)) =>
+      if (occurs(y, substTy1)) {
+        do raise(OccursCheckFailure(y, substTy1), "")
+      } else {
+        subst.put(y, substTy1)
+      }
+
+    // If both are constructors, unify their arguments
+    case (Con(name1, args1), Con(name2, args2)) =>
+      if (name1 != name2 || args1.size != args2.size) {
+        do raise(UnificationFailure(substTy1, substTy2), "Different number of arguments!")
+      } else {
+        unifyMany(args1, args2, subst)
+      }
+  }
+}
+
+// Unify a list of arguments with a current substitution
+def unifyMany(args1: List[Type], args2: List[Type], subst: Substitution): Substitution / Exception[UnificationError] =
+  (args1, args2) match {
+    case (Nil(), Nil()) => subst
+    case (Cons(a1, rest1), Cons(a2, rest2)) =>
+      val newSubst = unify(a1, a2, subst)
+      unifyMany(rest1, rest2, newSubst)
+    case _ => do raise(UnificationManyFailure(args1, args2), "Different numbers of types on each side!")
+  }
+
+def unify(ty1: Type, ty2: Type): Substitution / Exception[UnificationError] =
+  unify(ty1, ty2, map::empty(box bytearray::compareStringBytes))
+
+def showType(ty: Type): String = ty match {
+  case Var(name) => name
+  case Con(name, Nil()) => name
+  case Con(name, args) =>
+    name ++ "(" ++ args.map { t => showType(t) }.join(", ") ++ ")"
+}
+
+def show(err: UnificationError): String = err match {
+  case OccursCheckFailure(variable, ty) =>
+    "Occurs check failed: " ++ variable ++ " occurs in " ++ showType(ty)
+  case UnificationFailure(ty1, ty2) =>
+    "Cannot unify " ++ showType(ty1) ++ " with " ++ showType(ty2)
+    case UnificationManyFailure(tps1, tps2) =>
+    "Cannot unify " ++ tps2.map { showType }.join(", ") ++ " with " ++ tps1.map { showType }.join(", ")
+}
+
+def reporting { body : => Substitution / Exception[UnificationError] }: Unit = {
+  val res = result[Substitution, UnificationError] {
+    body()
+  }
+
+  res match {
+    case Success(subst) => {
+      println("Unification successful!")
+      println("Substitution:")
+      for { subst.each } { case (k, v) =>
+        println("  " ++ k ++ " -> " ++ showType(v))
+      }
```

Comment:
Funnily enough, this doesn't seem to be the culprit of non-termination: even if you replace it with a `subst.foreach { (k, v) =>`, it still hangs.